Adding documentation for android export (#2542)
o3de/o3de.org: content/docs/user-guide/packaging/project-export/project-export-android.md

---
linkTitle: Project Export for Android
title: Project Export for Android
description: Learn how to use the Project Export CLI to automate preparing your project for release on Android.
toc: true
weight: 410
---

> This guide covers how to use project export tooling to build and deploy games on Android. We will use the Android export script to generate a gradle project and prepare assets and then build the project. The end result of export should be an APK file that can be installed on your Android device, along with an additional step to auto-install on a connected phone.

Reviewer: ```suggestion
This guide covers how to use project export tooling to build and deploy games on Android. We will use the Android export script to generate a Gradle project, prepare assets, and then build the project. The end result of export should be an APK file that can be installed on your Android device, along with an additional step to auto-install on a connected phone.
```

4. Ensure that you have a proper keystore file configured for signing. You can learn more [here](https://development--o3deorg.netlify.app/docs/user-guide/platforms/android/#apk-signing). Here is an example keystore file you can create to start with (you will need the JDK with `keytool`):
```
export KEYSTORE_FILE_PATH=/path/to/android-key.keystore
export STOREPASS=o3depass
export ALIAS=o3dekey
export KEYPASS=o3depass
export KEYSIZE=2048
export VALIDITY=10000
export DNAME="cn=o3de-android-project, ou=o3de, o=LF, c=US"
keytool -genkey -keystore $KEYSTORE_FILE_PATH -storepass $STOREPASS  -alias $ALIAS -keypass $KEYPASS -keyalg RSA -keysize $KEYSIZE -validity $VALIDITY -dname $DNAME

> ```

Reviewer 2: `export` is used on Linux, while 'set' is used on windows to set environment variables. As this guide is focused on windows, this should at least use the windows `set` instead.

> 5. Make sure the o3de engine bootstrap registry is enabled for android. To do so, go to your O3DE engine installation and edit the file at `$O3DE_ENGINE_PATH/Registry/bootstrap.setreg` so that the `assets` field uses `android`. Like so:

Reviewer: ```suggestion
5. Make sure the O3DE bootstrap registry is enabled for Android. To do so, go to your O3DE installation and edit the file at `$O3DE_ENGINE_PATH/Registry/bootstrap.setreg` so that the `assets` field uses `android`. Like so:
```

```
"assets": "android",
```
6. Edit the `$O3DE_ENGINE_PATH/Registry/AssetProcessorPlatformConfig.setreg` file at line 67 by uncommenting the line:
```
//"android": "enabled", -> "android": "enabled",

> ```

Reviewer 2: I don't think this is necessary if you pass in the platform to asset processor batch anymore.

> This will tell the asset processor to make sure Android assets are cached when building the project.

Reviewer 2: ```suggestion
This will tell the asset processor to make sure assets are processed for Android when building the project.
```

> This will tell the asset processor to make sure ios assets are cached when building the project.

Reviewer: ```suggestion
This will tell the asset processor to make sure iOS assets are cached when building the project.
```

7. Make sure your project has the right configuration settings for Android. You can use the `android-configure` CLI tool to ensure this is setup. To see what is currently configured, run:
```
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure -l
```
You can also validate your current configuration using:
```
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --validate
```

> If you do not already have android configured, or encounter issues with the validate command, you can run the following:

Reviewer: ```suggestion
If you do not already have Android configured, or encounter issues with the validate command, you can run the following:
```

```
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --set-value platform.sdk.api=30
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --set-value ndk.version=25.*
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --set-value android.gradle.plugin=8.1.0
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --set-value sdk.root=/path/to/android/sdk
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --set-value signconfig.store.file=/path/to/android-key.keystore
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --set-value signconfig.key.alias=o3dekey
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --set-value asset.mode=PAK
```

> Note that for sdk.root and signconfig.store.file, you must already know where those things exist on your hard drive, and use those paths accordingly. To setup the keypass and storepass, you must use the following commands

Reviewer: ```suggestion
Note: you must already know where sdk.root and signconfig.store.file exist on your hard drive, and use those paths accordingly. To setup the keypass and storepass, you must use the following commands
```

```
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --set-password signconfig.store.password
$O3DE_ENGINE_PATH/scripts/o3de.bat android-configure --set-password signconfig.key.password
```
Make sure the passwords supplied here match what you used in the keystore file, and that you remember what they are.

## Quickstart
### Running the Export Script
For building in release mode with bundled PAK files for your APK, you can use the following export command:
```
export O3DE_ENGINE_PATH=/path/to/o3de

> export PROJECT_PATH=/path/to/project

Reviewer 2: same comment about using `set` instead

cd $PROJECT_PATH

> $O3DE_ENGINE_PATH/scripts/o3de.sh export-project -es export_source_android.py -pp . -abp build/game-android -ll INFO --config release --asset-mode PAK

Reviewer 2: The example above uses `.bat` which make this page appear windows specific. But this line switches to `o3de.sh` which is mac/linux specific. Suggest you use windows paths for the entire doc for consistency, especially if windows was the platform you tested these steps with.

```

To use profile mode instead, change the `--config` parameter in the above snippet to use `profile` instead. If you wish to set the asset mode to LOOSE, use `--asset-mode LOOSE`.

> That invocation should be all that is needed to create an APK file for your project. To install the APK on your phone for testing, this can be done by adding the `--deploy-to-android` flag in the export snippet above. Please note that to use this feature, you must ensure an android device is connected to your computer, and that ADB is able to properly access the device.

Reviewer: _"That invocation should be all that is needed to create an APK file for your project"_
😄 👍
